### DISCUSSION TITLE
Using compreface-python-sdk with Flask? Are there any secrets to improve performance?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [exadel-inc/compreface-python-sdk#14] [Q&A]
Are there any secrets on how to improve performance with Flask and gunicron?  

Should I place service connection strings in Flask __init__ portion, or is it better to define them in each function separately?
Won't it kill concurrency? (I mean that the workers will try to reuse global connection or something??)

```
compre_face: CompreFace = CompreFace(DOMAIN, PORT)

recognition: RecognitionService = compre_face.init_face_recognition(API_KEY)

face_collection: FaceCollection = recognition.get_face_collection()

```
What about using compreface-python-sdk with gunicorn? Are there any possible problems? Or everything should work as simple synchronous code? 

if I call compreface-python-sdk detect from several separate applications - will they be worked on simultaneously? 
Does compreface work on compreface-python-sd requests at the same time (using workers or threads) or does it use something like a queue and work on the tasks sequentially?

**Comment by pospielov** [comment]
It's better to init CompreFace SDK once. It uses `requests` package under the hood. So it should work concurrently if you run it from different processes.
CompreFace uwsgi starts with a single process. So it's better to update https://github.com/exadel-inc/CompreFace/blob/master/embedding-calculator/uwsgi.ini file and replace it with docker volume 
I'll create a task to make it easier to tweek

**Reply by martinenkoEduard** [reply]
Can you please share an example of uwsgi.ini that uses multiple threads? 

Does setting several "api_key string required Face Verification Api Key in UUID format" mean that i will increase concurrency?
Or is it completely irrelevant?

Why do we need different services and different api_keys?

**Reply by martinenkoEduard** [reply]
Should I increase processes =  or threads =  in uwsgi.ini file?

**Reply by martinenkoEduard** [reply]
Should I increase something else? Like Java configurations files?

Will increasing processes = or threads = in uwsgi.ini spawn more processes/models ? How much more memory will it consume?

**Reply by martinenkoEduard** [reply]
Doesn't new api_key creates new thread?

**Reply by pospielov** [reply]
Sorry for misleading you.
The default number of processes is 2 for CPU builds and 1 for GPU
You can change it by setting the environment variable UWSGI_PROCESSES for `compreface-core` container

> Why do we need different services and different api_keys?

API keys are not related to threads in any way. It makes sense to have different face recognition services because they have their own face collections. For face detection and face verification services, the advantages are not so clear. The main advantage is improved security - you can delete the service if someone stole your API key.

> Will increasing processes = or threads = in uwsgi.ini spawn more processes/models ? How much more memory will it consume?

By increasing the number of processes in uwsgi, you increase your memory footprint. As processes do not share a memory, each process loads all neural networks in memory.

**Comment by martinenkoEduard** [comment]
And does detection and recognition services work independently of each other?  (so can simultaneously server clients?)

**Reply by pospielov** [reply]
It doesn't matter if you use `detection and recognition services`, or the one recognition service. The logic of handling requests is the same.

**Reply by martinenkoEduard** [reply]
Btw - can you explain in simple term how it works - 
Do you use django as REST API? But JAVA models? How does django spawn a new java model? 
How are models and REST API connected?

**Reply by pospielov** [reply]
Here is the description of our servers:
https://github.com/exadel-inc/CompreFace/blob/master/docs/Architecture-and-scalability.md
In short, Python runs neural networks, Java is used for validation and calculating similarity. 
They communicate using REST API.
